### PR TITLE
fix(errors): correct the client-boundary example in the error catalog

### LINKED ISSUE #4050: Malformed client-boundary example in the user-facing error catalog
## What

The `client-boundary-violation` entry in `src/errors/user-friendly/error-catalog.ts:61-74` ships a malformed code example. This string is shown to users when they hit a client-boundary error, so it is product copy, not an internal comment.

Current value:

```
// ❌ Wrong - database import in client component
'use client';
import { db } from './database'; // Error!

import { db } from './database';
export default async function ServerComponent() {
  const data = await db.query('...');
  return <ClientComponent data={data} />;
}

'use client';
export default function ClientComponent({ data }) {
}
```

## Problems

1. **The corrected half carries no label.** Only the wrong half is marked `// ❌ Wrong`. Everything after the blank line is the fix, but nothing says so, so the whole block reads as one broken snippet.
2. **`import { db } from './database'` appears twice**, once in the wrong half and once in the fix, with no file boundary between them. A reader cannot tell these are meant to be separate files.
3. **The second `'use client'` sits after the imports**, where the directive is inert. A directive must be the first statement in the file. As written, the example demonstrates the exact mistake it is meant to correct.
4. **`ClientComponent` has an empty body**, so it renders nothing and never uses the `data` prop the example is built around.

## Why it matters

`AGENTS.md` requires that code examples be complete, copyable, and safe to paste. A user who copies this gets a component that renders nothing and a `'use client'` directive that does not apply.

## Suggested fix

Split the example into two clearly labeled files, put each `'use client'` on the first line of its file, and give `ClientComponent` a body that consumes `data`.

## Provenance

Found during a test-quality audit of `src/errors` (PR #4049). Raised separately rather than fixed there because correcting it means rewriting user-facing copy, which is outside the scope of a test PR. No test asserted the example's structure; the new test added in #4049 covers the entry's shape, which is why the malformed content passed.

### PR DESCRIPTION
**Pull request by kojiwakayama**
## Description

The `client-boundary` entry ships a malformed code example. It is shown to users when they hit a client-boundary error, so it is product copy rather than an internal comment, and `AGENTS.md` requires examples to be complete, copyable and safe to paste.

All four defects from the issue are fixed:

1. **The corrected half carried no label.** Only the wrong half was marked, so everything after the blank line read as part of one broken snippet. Both halves are now marked.
2. **`import { db } from './database'` appeared twice with no file boundary.** The example now uses the `// app/...` file-path comment convention the other entries in this catalog already follow, so the three blocks are visibly separate files.
3. **The second `'use client'` sat after the imports**, where the directive is inert — the example demonstrated the exact mistake it exists to correct. Each directive is now the first line of its own file.
4. **`ClientComponent` had an empty body**, so it rendered nothing and never used the `data` prop the example is built around. It now renders the data.

## Related Issue(s)

Fixes #4050

## Type of Change

- [x] Documentation update

## Checklist

- [x] I have added tests that prove my fix is effective

## Testing

The issue notes: *"No test asserted the example's structure; the new test added in #4049 covers the entry's shape, which is why the malformed content passed."* So the fix is only half the work — three assertions now cover the **structure of the copy**, across every catalog entry that has an example:

- a `'use client'` directive must be the first statement of its file block
- an example that marks a wrong form must mark the corrected one too
- the client-boundary example must consume the data it threads across the boundary

Verified against the original catalog: all three fail. One of them only fails because I fixed the test — my first version used `indexOf`, which finds the *earliest* directive and so passed on the old example, whose defect was the **second** directive appearing after the imports. It now checks every occurrence.

```
deno task test:file src/errors/     32 passed (523 steps) | 0 failed
deno task docs:errors:check         docs/guides/errors.md is current (122 errors)
deno task docs:api-reference:check / lint:anti-slop / lint:test-typecheck
deno fmt --check, deno lint
```

## Summary by CodeRabbit

* **Documentation**
  * Updated client/server component boundary examples with clearer ❌ invalid and ✅ corrected patterns.
  * Added a typed database query example that passes selected user data to a client component.
* **Tests**
  * Expanded validation for client directives, component boundaries, data passing, comments, strings, semicolons, and directive edge cases.
  * Updated examples to verify correct directive placement and server-to-client prop usage.

### CODE REVIEW

File: src/errors/user-friendly/error-catalog.ts [hunk 1]
```diff
@@ -58,18 +58,28 @@ export async function POST(request: Request) {
...
 
+// ✅ Correct: app/dashboard/page.tsx (Server Component, no directive)
 import { db } from './database';
-export default async function ServerComponent() {
-  const data = await db.query('...');
-  return <ClientComponent data={data} />;
+import { UserList } from './user-list';
+
+export default async function DashboardPage() {
+  const users = await db.query('SELECT name FROM users');
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.ts]: Verified against the contract and you are right. Fixed in a1acdf988.

`src/extensions/database/database-client.ts`:

```ts
export interface QueryResult<T = Record<string, unknown>> {
  rows: T[];
  rowCount: number;
}
query<T>(sql: string, params?: unknown[]): Promise<QueryResult<T>>;
```

So `const users = await db.query(...)` bound the whole result and `users.map(...)` would have thrown at render. For an example whose entire purpose is to show the working remediation, that is the worst place to be wrong — worse than the original malformed copy, because it looks correct.

Now `const { rows } = await db.query<{ name: string }>(...)` with `rows` passed across the boundary.

File: src/errors/user-friendly/error-catalog.ts [hunk 2]
```diff
@@ -58,18 +58,28 @@ export async function POST(request: Request) {
...
+  return <UserList users={users} />;
 }
 
+// ✅ Correct: app/dashboard/user-list.tsx (Client Component)
 'use client';
-export default function ClientComponent({ data }) {
+
+export function UserList({ users }) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.ts]: Agreed, fixed in a1acdf988.

`AGENTS.md` requires examples to be complete and safe to paste, and a `.tsx` file with an implicitly-typed destructured prop does not compile under strict TypeScript, which is the setting a Veryfront project would actually use.

Both sides now carry a structural row type, and the query is generic so its rows match the prop:

```tsx
const { rows } = await db.query<{ name: string }>('SELECT name FROM users');
return <UserList users={rows} />;

export function UserList({ users }: { users: { name: string }[] }) { ... }
```

Structural rather than a shared exported type, so each file stays independently copyable — which is the property the example needs.

File: src/errors/user-friendly/error-catalog.test.ts [hunk 3]
```diff
@@ -110,6 +110,66 @@ describe("ERROR_SOLUTIONS", () => {
...
+    // paste. These assert the shape of the copy rather than its wording, which
+    // is what the entry-shape tests could not catch.
+    it("places every 'use client' directive at the top of its own file block", () => {
+      for (const [key, solution] of Object.entries(ERROR_SOLUTIONS)) {
+        const example = solution.example;
+        if (!example?.includes("'use client'")) continue;
+
+        // A file block starts at a leading `// ...:` path comment. Within one,
+        // the directive must precede any statement, or it is inert.
+        for (const block of example.split(/\n(?=\/\/ )/)) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Good catch on the guard rather than the code, and it was a real hole. Fixed in a1acdf988.

Splitting on any full-line comment meant an ordinary comment could open a block, so a misplaced directive would land at index 0 of it and pass. I ran your counter-example against both forms:

```
// ✅ Correct: app/x.tsx (Client Component)
import { a } from './a';
// a plain explanatory comment
'use client';
```

```
loose  "// "    misplaced directive detected = False
strict ❌/✅     misplaced directive detected = True
```

So the guard would have passed exactly the defect it exists to catch. It now splits only on the documented header form, `// ❌ Wrong: path` or `// ✅ Correct: path`.

This is the second time on this PR that my *test* was the weak part rather than the fix — the first was using `indexOf`, which found only the earliest directive. Both are now covered.

File: src/errors/user-friendly/error-catalog.test.ts [hunk 4]
```diff
@@ -110,6 +110,68 @@ describe("ERROR_SOLUTIONS", () => {
...
+    it("places every 'use client' directive at the top of its own file block", () => {
+      for (const [key, solution] of Object.entries(ERROR_SOLUTIONS)) {
+        const example = solution.example;
+        if (!example?.includes("'use client'")) continue;
+
+        // Split only on the documented file-header form, `// ❌ Wrong: path` or
+        // `// ✅ Correct: path`. Splitting on any full-line comment would let an
+        // ordinary comment start a new block and report a misplaced directive
+        // as the first statement of it, which passes invalid code.
+        for (const block of example.split(/\n(?=\/\/ (?:❌|✅))/)) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Right again, and this is the third iteration on this one guard. Fixed in 060612e59.

My previous tightening (any comment starting with ❌ or ✅) still lets an ordinary annotation open a block. Ran your counter-example against both:

```
// ✅ Correct: app/x.tsx (Client Component)
import { a } from './a';
// ✅ data is validated
'use client';
```

```
loose ❌/✅       misplaced directive detected = False
strict header    misplaced directive detected = True
```

Splitting now requires the complete header form, `// ❌ Wrong: path` or `// ✅ Correct: path`.

Worth naming the pattern, since it is three for three: each time I tightened this guard I tightened it just enough for the case in front of me. The general shape is that the guard has to recognise the *documented convention*, not an approximation of it, or it keeps admitting whatever sits outside the approximation.

File: src/errors/user-friendly/error-catalog.ts [hunk 5]
```diff
@@ -58,18 +58,28 @@ export async function POST(request: Request) {
       "Use 'use server' directive for server actions",
       "Split component into server and client parts",
     ],
-    example: `// ❌ Wrong - database import in client component
+    example: `// ❌ Wrong: app/dashboard/page.tsx
 'use client';
-import { db } from './database'; // Error!
+import { db } from './database'; // A Client Component cannot reach the database
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.ts:63]: Agreed, fixed in 060612e59.

The wrong block was a directive plus an import and nothing else, so a reader taking it as the standalone `app/dashboard/page.tsx` it is labelled as would have a file that exports no page. That undercuts the label.

It now exports a page, and is wrong in exactly the one way the entry is about — a Client Component reaching the database — rather than being incomplete as well as wrong:

```tsx
// ❌ Wrong: app/dashboard/page.tsx
'use client';
import { db } from './database'; // A Client Component cannot reach the database

export default async function DashboardPage() {
  const { rows } = await db.query<{ id: number; name: string }>('SELECT id, name FROM users');
  return <ul>{rows.map((user) => <li key={user.id}>{user.name}</li>)}</ul>;
}
```

File: src/errors/user-friendly/error-catalog.ts [hunk 6]
```diff
@@ -58,18 +58,28 @@ export async function POST(request: Request) {
...
 }
 
+// ✅ Correct: app/dashboard/user-list.tsx (Client Component)
 'use client';
-export default function ClientComponent({ data }) {
+
+export function UserList({ users }: { users: { name: string }[] }) {
+  return (
+    <ul>
+      {users.map((user) => <li key={user.name}>{user.name}</li>)}
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.ts]: Fixed in 060612e59. Both queries now select `id` and both lists key by it.

Keying on `name` collides as soon as two users share one, and React would reuse the wrong element on reorder. For copy that exists to be pasted, the example should teach the habit rather than the shortcut, so the id is selected explicitly rather than falling back to the array index.

The prop type moved with it, so the server and client files still match structurally and each stays independently copyable.

Added an assertion pinning it, so a future edit cannot quietly reintroduce `key={user.name}`.

File: src/errors/user-friendly/error-catalog.test.ts [hunk 7]
```diff
@@ -110,6 +110,72 @@ describe("ERROR_SOLUTIONS", () => {
...
+          const statements = block
+            .split("\n")
+            .map((line) => line.trim())
+            .filter((line) => line !== "" && !line.startsWith("//"));
+
+          // Every occurrence must be first, not just the earliest one: the
+          // defect this guards against was a *second* directive sitting after
+          // the imports of a block, where it is inert.
+          const directiveIndexes = statements
+            .map((statement, index) => (statement === "'use client';" ? index : -1))
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Valid — the exact-equality comparison against `"'use client';"` recorded no indexes for the semicolonless form (or the double-quoted one), so a misplaced directive in those spellings passed the guard. Fixed in 9a80fbea8: the guard now matches the directive syntax, `/^['"]use client['"];?$/`, and the engagement filter accepts any quote style too. Verified the semicolonless and double-quoted misplaced counter-examples are both detected and a well-placed semicolonless directive is not flagged.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/errors/user-friendly/error-catalog.test.ts [hunk 8]
```diff
@@ -110,6 +110,77 @@ describe("ERROR_SOLUTIONS", () => {
...
 
+  describe("example structure", () => {
+    // AGENTS.md requires code examples to be complete, copyable and safe to
+    // paste. These assert the shape of the copy rather than its wording, which
+    // is what the entry-shape tests could not catch.
+    it("places every 'use client' directive at the top of its own file block", () => {
+      // Match the directive syntax, not one exact spelling: single or double
+      // quotes, with or without the trailing semicolon, are all forms
+      // Veryfront honours and project templates use.
+      const directive = /^['"]use client['"];?$/;
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Valid, and it settles the question of approach — this was the fourth hole with the same shape, so the fix in 169de5b23 stops patching the regex and parses the directive prologue instead.

The guard now judges placement on statements, not raw lines: a small string-aware scanner strips `//` and `/* */` comments (preserving line structure, and leaving string contents like `'https://x'` intact), the remaining trimmed non-empty lines are the block's statements, and every `'use client'`/`"use client"` expression statement — semicolon optional — must be statement 0 of its file block. A trailing `// Client Component` is stripped before matching, so the anchored form can no longer hide a misplaced directive, and a commented-out directive is no longer counted as one.

The parse itself is now pinned by direct test cases, so the catalog test cannot silently stop enforcing: a misplaced directive with a trailing line comment is detected, a comment-and-blank-line prologue does not displace a leading directive, a directive after a statement is flagged, and a directive inside a line or block comment is ignored.

`deno test` on the file passes (24 steps), `deno fmt --check`, `deno lint`, and `deno check` on the file are clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/errors/user-friendly/error-catalog.ts [hunk 9]
```diff
@@ -58,18 +58,33 @@ export async function POST(request: Request) {
...
+
+// ✅ Correct: app/dashboard/page.tsx (Server Component, no directive)
 import { db } from './database';
-export default async function ServerComponent() {
-  const data = await db.query('...');
-  return <ClientComponent data={data} />;
+import { UserList } from './user-list';
+
+export default async function DashboardPage() {
+  const { rows } = await db.query<{ id: number; name: string }>('SELECT id, name FROM users');
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.ts:74]: Valid - the canonical entry is what `getErrorSolution` and the dev dashboard actually serve, so the correction had to land there too. Fixed in 226978f42: the `RSC_ERROR_CATALOG["client-boundary-violation"]` example now destructures `rows` from the typed `db.query<{ id: number; name: string }>(...)` result, passes `rows` across the boundary, and uses a structural prop type instead of importing `User` from an undefined `./types.ts`, so both public surfaces now serve the same copyable remediation as the corrected `ERROR_SOLUTIONS` entry. The rsc-errors test now pins the destructured rows and the `data={rows}` boundary pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/errors/user-friendly/error-catalog.test.ts [hunk 10]
```diff
@@ -3,6 +3,90 @@ import { assert, assertEquals, assertExists } from "#veryfront/testing/assert.ts
...
+const directive = /^['"]use client['"];?$/;
+
+// The statement lines of a file block, comments stripped first: a trailing
+// comment cannot disguise a directive, and a commented-out directive is not
+// mistaken for a real one.
+function statementsOf(block: string): string[] {
+  return stripComments(block)
+    .split("\n")
+    .map((line) => line.trim())
+    .filter((line) => line !== "");
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Addressed in the current head. The guard now strips comments and splits code-level semicolons while remaining string-aware; its regression verifies that an import and use client on one line are parsed as separate statements and classified as misplaced.

File: src/errors/user-friendly/error-catalog.test.ts [hunk 11]
```diff
@@ -110,6 +202,155 @@ describe("ERROR_SOLUTIONS", () => {
...
+        for (const block of example.split(/\n(?=\/\/ (?:❌ Wrong|✅ Correct): )/)) {
+          const statements = statementsOf(block);
+
+          // Every occurrence must be first, not just the earliest one: the
+          // defect this guards against was a *second* directive sitting after
+          // the imports of a block, where it is inert.
+          for (const index of useClientIndexes(block)) {
+            assertEquals(
+              index,
+              0,
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Fixed in 7b548b408. The guard now permits preceding string-literal directives and rejects use client only when a non-directive statement has already ended the prologue. Added coverage for use strict and use client sharing one line.

File: src/errors/user-friendly/error-catalog.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.ts:74]: Addressed in the current head. The canonical RSC catalog now destructures rows from the typed query result and passes rows across the boundary, with focused assertions in rsc-errors.test.ts. The user-friendly catalog uses the same copyable shape.

**Comment by kojiwakayama** [comment]
@codex review

File: src/errors/user-friendly/error-catalog.test.ts [hunk 12]
```diff
@@ -3,6 +3,105 @@ import { assert, assertEquals, assertExists } from "#veryfront/testing/assert.ts
...
+// project templates use.
+const directive = /^['"]use client['"];?$/;
+
+// The statements of a file block, comments stripped and semicolons resolved
+// first: a trailing comment cannot disguise a directive, a commented-out
+// directive is not mistaken for a real one, and a directive sharing a line
+// with an earlier statement is still seen in its true position.
+function statementsOf(block: string): string[] {
+  return toStatementSource(block)
+    .split("\n")
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Fixed in b9deef4e4. The guard now uses the runtime hasUseClientDirective classifier as the source of truth for whether a leading string is actually in the directive prologue, while retaining the per-occurrence scan that catches later inert duplicates. Added the exact ASI continuation regression; it failed before the fix and now passes. The full errors suite, lint:ci, and typecheck all pass.

**Comment by kojiwakayama** [comment]
@codex review

Please review exact head b9deef4e4. The ASI continuation finding is fixed and resolved. The focused regression failed before the fix; the full errors suite, lint:ci, and typecheck pass.

File: src/errors/user-friendly/error-catalog.test.ts [hunk 13]
```diff
@@ -1,8 +1,114 @@
...
+
+const stringLiteralStatement = /^(?:'(?:\\.|[^'\\])*'|"(?:\\.|[^"\\])*")$/;
+
+function invalidUseClientIndexes(block: string): number[] {
+  const statements = statementsOf(block);
+  const indexes = useClientIndexes(block);
+  // Use the runtime classifier as the source of truth for whether a leading
+  // string is actually a directive. In particular, a semicolonless string
+  // followed by an ASI continuation token is one expression, not a directive.
+  if (indexes.length > 0 && !hasUseClientDirective(block)) return indexes;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Addressed in 261f88fdf. The guard now records every comment-stripped statement containing a directive-like `'use client'` spelling and invokes the runtime classifier whenever one appears. If the runtime says the block is not a Client Component, those occurrences fail the catalog check, including the new `('use client');` regression.

**Comment by kojiwakayama** [comment]
@codex review

File: src/errors/user-friendly/error-catalog.ts [hunk 14]
```diff
@@ -58,18 +58,33 @@ export async function POST(request: Request) {
       "Use 'use server' directive for server actions",
       "Split component into server and client parts",
     ],
-    example: `// ❌ Wrong - database import in client component
+    example: `// ❌ Wrong: app/dashboard/page.tsx
 'use client';
-import { db } from './database'; // Error!
+import { db } from './database'; // A Client Component cannot reach the database
 
+export default async function DashboardPage() {
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/errors/user-friendly/error-catalog.ts]: Addressed in c5dba0161. The intentionally wrong Client Component now exports a synchronous DashboardPage. It still imports and invokes the server-only database client, so the example isolates the intended client/server boundary violation without adding an async Client Component failure. The focused catalog test now pins this property.

File: src/errors/user-friendly/error-catalog.test.ts [hunk 15]
```diff
@@ -1,8 +1,121 @@
...
+    const char = source[i]!;
+    const next = source[i + 1];
+
+    if (state === "code") {
+      if (char === "/" && next === "/") {
+        state = "line-comment";
+        i++;
+      } else if (char === "/" && next === "*") {
+        state = "block-comment";
+        i++;
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Addressed in c5dba0161. The hand-written comment scanner has been replaced with the repository pinned Babel parser in TypeScript and JSX mode. Top-level directives and statements are ordered from the parsed AST, so /[/*]/ remains a regex literal and the later inert use-client statement is detected. The exact counterexample is covered by a regression test.

**Comment by kojiwakayama** [comment]
@codex review

File: src/errors/user-friendly/error-catalog.test.ts [hunk 16]
```diff
@@ -1,8 +1,65 @@
...
+  const indexes = useClientIndexes(block);
+  const directiveLikeIndexes = statements
+    .map((statement, index) => (directiveLike.test(statement) ? index : -1))
+    .filter((index) => index !== -1);
+  // Use the runtime classifier as the source of truth for whether a leading
+  // string is actually a directive. In particular, a semicolonless string
+  // followed by an ASI continuation token or a wrapped string is one
+  // expression, not a directive.
+  if (directiveLikeIndexes.length > 0 && !hasUseClientDirective(block)) {
+    return directiveLikeIndexes;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.test.ts]: Fixed in cac294eb5. The example validator now classifies Babel directive nodes and expression statements structurally. Bare and wrapped use-client literals, plus ASI continuations that start from the literal, remain covered; directive spelling inside a throw message or other statement is ignored. Verified with 31 user-friendly catalog steps, 6 RSC catalog steps, typecheck, generated docs checks, and full lint:ci.

**Comment by kojiwakayama** [comment]
@codex review

**Comment by kojiwakayama** [comment]
@codex review

File: src/errors/user-friendly/error-catalog.test.ts [hunk 17]
```diff
@@ -1,8 +1,119 @@
...
+      continue;
+    }
+    if (
+      expression.type === "ParenthesizedExpression" || expression.type === "TSAsExpression" ||
+      expression.type === "TSTypeAssertion" || expression.type === "TSNonNullExpression"
+    ) {
+      expression = expression.expression;
+      continue;
+    }
+    return false;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/errors/user-friendly/error-catalog.test.ts:105]: Fixed at exact PR head 211e61d0dacd256f20d84990592209c1395e9889. The guard now follows binary/logical left operands, conditional tests, sequence heads, tagged-template tags, and TypeScript wrappers; the regression table covers +, &&, ?:, comma, and tagged-template ASI continuations. Focused catalog and runtime-classifier tests pass, and the full pre-push suite passed 4,306 tests.

**Comment by kojiwakayama** [comment]
@codex review

**Comment by kojiwakayama** [comment]
@codex review
